test/e2e/split/main.go

```diff
@@ -0,0 +1,225 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"math"
+	"os"
+	"path/filepath"
+	"regexp"
+	"sort"
+	"strings"
+)
+
+type options struct {
+	numChunks  int
+	printChunk int
+	printDebug bool
+	writer     io.Writer
+}
+
+func main() {
+	opts := options{
+		writer: os.Stdout,
+	}
+	flag.IntVar(&opts.numChunks, "chunks", 1, "Number of chunks to create focus regexps for")
+	flag.IntVar(&opts.printChunk, "print-chunk", 0, "Chunk to print a regexp for")
+	flag.BoolVar(&opts.printDebug, "print-debug", false, "Print all spec prefixes in non-regexp format. Use for debugging")
+	flag.Parse()
+
+	dir := flag.Arg(0)
+	if dir == "" {
+		exitIfErr(fmt.Errorf("test directory required as the argument"))
+	}
+
+	// Clean dir.
+	var err error
+	dir, err = filepath.Abs(dir)
+	exitIfErr(err)
+	wd, err := os.Getwd()
+	exitIfErr(err)
+	dir, err = filepath.Rel(wd, dir)
+	exitIfErr(err)
+
+	exitIfErr(opts.run(dir))
+}
+
+func exitIfErr(err error) {
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+func (opts options) run(dir string) error {
+	describes, err := findDescribes(dir)
+	if err != nil {
+		return err
+	}
+
+	// Find minimal prefixes for all spec strings so no spec runs are duplicated across chunks.
+	prefixes := findMinimalWordPrefixes(describes)
+	sort.Strings(prefixes)
+
+	var out string
+	if opts.printDebug {
+		out = strings.Join(prefixes, "\n")
+	} else {
+		out, err = createChunkRegexp(opts.numChunks, opts.printChunk, prefixes)
+		if err != nil {
+			return err
+		}
+	}
+
+	fmt.Fprint(opts.writer, out)
+	return nil
+}
+
+// TODO: this is hacky because top-level tests may be defined elsewise.
+// A better strategy would be to use the output of `ginkgo -noColor -dryRun`
+// like https://github.com/operator-framework/operator-lifecycle-manager/pull/1476 does.
+var topDescribeRE = regexp.MustCompile(`var _ = Describe\("(.+)", func\(.*`)
+
+func findDescribes(dir string) ([]string, error) {
+	// Find all Ginkgo specs in dir's test files.
+	// These can be grouped independently.
+	describeTable := make(map[string]struct{})
+	matches, err := filepath.Glob(filepath.Join(dir, "*_test.go"))
+	if err != nil {
+		return nil, err
+	}
+	for _, match := range matches {
+		b, err := ioutil.ReadFile(match)
+		if err != nil {
+			return nil, err
+		}
+		specNames := topDescribeRE.FindAllSubmatch(b, -1)
+		if len(specNames) == 0 {
+			log.Printf("%s: found no top level describes, skipping", match)
```

Comment:
It would be nice to control the log verbosity of this but I'm not too concerned if we're namely running this package in CI environments.

Reply:
That's what I was thinking too. I'll leave a note for a follow up.